[2단계 - 점심 뭐 먹지] 분주(황현정) 미션 제출합니다.

bunju20 opened the pull request:
안녕하세요 온스타🌟 이번에도 리뷰 잘부탁드립니다! 이번에는 2단계를 개발했는데요. 특정 기능이 추가되고, 도메인부분을 ts로 마이그레이션 하는것이 요구사항이었습니다! 폴더구조는 다음과 같아요!

```md
📦src
 ┣ 📂components - UI 컴포넌트들
 ┃ ┣ 📜AddDetailModal.js - 레스토랑 상세 정보 모달
 ┃ ┣ 📜AddRestaurantModal.js - 새 레스토랑 추가 모달
 ┃ ┣ 📜RestaurantFilterContainer.js - 필터링 UI 컴포넌트
 ┃ ┣ 📜RestaurantItem.js - 개별 레스토랑 아이템 컴포넌트
 ┃ ┣ 📜RestaurantList.js - 레스토랑 목록 컴포넌트
 ┃ ┗ 📜RestaurantTabs.js - 자주가는 음식점 탭 컴포넌트
 ┣ 📂constants - 상수 정의
 ┃ ┣ 📜errors.js - 오류 메시지 상수
 ┃ ┣ 📜messages.js - 일반 메시지 상수
 ┃ ┗ 📜options.js - 옵션 관련 상수
 ┣ 📂data - 데이터 모델 (TypeScript 적용됨)
 ┃ ┗ 📜initialRestaurants.ts - 레스토랑 데이터 모델
 ┣ 📂handlers - 이벤트 핸들러 (TypeScript 적용됨)
 ┃ ┣ 📜detailModalHandler.ts - 상세 모달 관련 이벤트 처리
 ┃ ┣ 📜favoriteHandler.ts - 즐겨찾기 기능 이벤트 처리
 ┃ ┣ 📜filterHandler.ts - 필터링 관련 이벤트 처리
 ┃ ┣ 📜modalHandler.ts - 모달 관련 공통 이벤트 처리
 ┃ ┣ 📜restaurantHandler.ts - 레스토랑 CRUD 이벤트 처리
 ┃ ┗ 📜tabHandler.ts - 자주가는 음식점 탭 전환 이벤트 처리
 ┣ 📂initialization - 앱 초기화 모듈
 ┃ ┣ 📜filterInitializer.js - 필터 컴포넌트 초기화
 ┃ ┣ 📜modalInitializer.js - 모달 기능 초기화
 ┃ ┣ 📜restaurantInitializer.js - 레스토랑 목록 초기화
 ┃ ┗ 📜tabInitializer.js - 탭 컴포넌트 초기화
 ┣ 📂shared - 공통 UI 컴포넌트
 ┃ ┣ 📜CustomButton.js - 버튼 컴포넌트
 ┃ ┣ 📜CustomDropdown.js - 드롭다운 컴포넌트
 ┃ ┗ 📜CustomInput.js - 입력 필드 컴포넌트
 ┣ 📂utils - 유틸리티 함수 (일부 TypeScript 적용됨)
 ┃ ┣ 📜categoryMapping.ts - 카테고리 매핑 유틸리티
 ┃ ┣ 📜generateId.js - ID 생성 유틸리티
 ┃ ┣ 📜localStorage.ts - 로컬 스토리지 관리 유틸리티
 ┃ ┗ 📜removeModal.js - 모달 제거 유틸리티
 ┣ 📂validation - 입력 유효성 검사
 ┃ ┗ 📜validator.js - 폼 입력값 검증 함수들
 ┗ 📜main.js - 애플리케이션 진입점
```

```mermaid
graph TD
    A[main.js] --> B[initialization]
    B --> C[components]
    B --> D[handlers]
    D --> E[utils]
    D --> F[data]
    C --> G[shared]
    D --> H[validation]
    F --> E
    
    subgraph "도메인 영역 (TypeScript)"
        F[data/initialRestaurants.ts]
        E2[utils/*.ts]
        D2[handlers/*.ts]
    end
    
    subgraph "UI 영역"
        C[components]
        G[shared]
        B[initialization]
    end
    
    classDef ts fill:#f9f,stroke:#333,stroke-width:2px;
    class F,E2,D2 ts;
```

```mermaid
sequenceDiagram
    participant User
    participant UI as 컴포넌트
    participant Handler as 이벤트 핸들러
    participant Data as 데이터 모델
    participant Storage as localStorage

    User->>UI: 사용자 입력 (필터링, 즐겨찾기 등)
    UI->>Handler: 이벤트 발생
    Handler->>Data: 데이터 업데이트
    Handler->>Storage: 데이터 저장
    Handler->>UI: UI 업데이트
    
    Note over User,UI: 레스토랑 클릭 시
    User->>UI: 레스토랑 항목 클릭
    UI->>Handler: detailModalHandler 호출
    Handler->>UI: 상세 모달 표시
    
    Note over User,UI: 자주가는 음식점 탭 선택 시
    User->>UI: 자주가는 음식점 탭 클릭
    UI->>Handler: tabHandler 호출
    Handler->>Data: 즐겨찾기 데이터 조회
    Handler->>UI: 즐겨찾기 레스토랑만 표시
```

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- 어플리케이션을 컴포넌트 단위로 모듈화하여 개발
- UI를 컴포넌트 단위로 생각하고 개발하는 연습
- 재사용할 수 있는 컴포넌트를 고민해보기
- TypeScript의 기본 문법을 익히며 필요성을 경험

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://bunju20.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용
이번단계에서는 1.리팩토링 2.기능 구현 3.ts로 마이그레이션과 같은 순서로 개발을 진행했습니다. 

**이벤트 리스너 관리**
하면서 이벤트버블링과 이벤트리스너 누락으로 인한 오류가 잦았습니다. 예를 들면 레스토랑을 눌러서 상세모달이 띄워진후에는 특정 버튼들의 이벤트 리스너가 없어졌다던지... 같은문제로요! 그래서 매번 리스너를 다시 붙여주는 식으로 구현을 했는데 이게 맞는지 잘 모르겠습니다. 비효율적이지 않나?? 하는 생각을 하면서 구현한것 같아요. 좀더 좋은 방식이 있나요...!?

**함수형 vs 클래스 기반 설계**
사실 함수형으로 프로젝트를 진행해본게 이번이 처음인데, 개발하면서 불편함을 많이 느낀것같습니다! 상태 관리와 이벤트 핸들러 관리 측면에서 클래스 기반으로 설계했다면 더 나았을지 궁금합니다. 현재 구조에서는 전역 변수와 함수들이 여러 파일에 분산되어 있다고 느껴졌던적이 많았던것같아요!

**DOM 조작과 상태 동기화**
UI 상태와 데이터 상태를 동기화하는 부분에서도 어려움이 있었는데요... 예를 들어 즐겨찾기를 모달에서 변경하면 목록에는 반영되지 않는 문제가 있어서 즐겨찾기 버튼 클릭 이벤트로 레스토랑 목록을 다시 렌더링하는식으로 구현했는데... 너무 비효율적인것처럼 느껴졌습니다!  더 좋은 방법이있을까요??

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

bunju20 committed:
feat: 브랜치생성

eunwoo-levi committed:
docs: 기능 목록 작성

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

bunju20 committed:
Merge branch 'step1' of https://github.com/eunwoo-levi/javascript-lunch into step1

bunju20 committed:
docs: 테스팅 목록 추가

eunwoo-levi committed:
feat: html, css 초기 세팅

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

eunwoo-levi committed:
chore: 필요없는 파일 삭제

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

eunwoo-levi committed:
feat: Modal 취소 기능 구현

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

eunwoo-levi committed:
docs: 취소하기 내용 업데이트

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

bunju20 committed:
test: 헤더 테스트 코드 추가
- 헤더 타이틀 테스트
- 헤더 모달 버튼 테스트

eunwoo-levi committed:
test: Body 테스트 코드 추가
- 가게 아이콘이 표시되는지 확인
- 가게 이름이 표시되는지 확인
- 캠퍼스 소요 시간이 표시되는지 확인
- 가게 설명이 표시되는지 확인

bunju20 committed:
test: 모달 테스팅 추가
- 배경 테스트
- 모달 제목 테스트
- 카테고리 드롭다운 테스트
- 이름 테스트

eunwoo-levi committed:
test: 모달 테스트 코드 작성
- 거리 클릭 시 항목들 표시 확인
- 설명 입력 확인
- 참고 링크 입력 확인
- 모달 하단에 취소하기, 추가하기 버튼 표시 확인

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

bunju20 committed:
feat: 추가하기 버튼 눌렀을때 가게 추가되는 기능 구현
- 추후 기능 수정 필요

eunwoo-levi committed:
feat: 새로운 레스토랑 추가하기 기능 구현

eunwoo-levi committed:
test: 추가하기 기능 테스트

bunju20 committed:
refactor: 헤더 컴포넌트 분리

eunwoo-levi committed:
refactor: 클래스형을 함수형으로 리팩토링

eunwoo-levi committed:
refactor: 레스토랑 리스트 분리

bunju20 committed:
refactor: restaurantList 분리

eunwoo-levi committed:
refactor: 모달 분리 리팩토링

bunju20 committed:
refactor: dropdown 컴포넌트 분리

eunwoo-levi committed:
refactor: Custom Input Element 분리

bunju20 committed:
refactor: CustomButton 분리 및 취소 버튼 수정

eunwoo-levi committed:
test: restaurant 이름 테스트코드 구현
- 이름 input 유효성 검사 로직 구현

bunju20 committed:
test: 설명 입력 검증 테스트 구현

bunju20 committed:
fix: 참고링크 maxLength 지정

eunwoo-levi committed:
test: 이름과 설명 입력 단위 테스트 구현

bunju20 committed:
feat: 에러 발생시 alert 출력 기능 구현

bunju20 committed:
refactor : select 테스트 코드 추가 및 개선

eunwoo-levi committed:
refactor: 카테고리 매핑함수 분리

Co-authored-by: bunju20 <bunju20@users.noreply.github.com>

bunju20 committed:
refactor : 레스토랑 필터 공통 컴포넌트 활용 리팩토링

bunju20 committed:
refactor: 이벤트 등록 전 돔 존재여부 확인 로직추가

bunju20 committed:
refactor: 헤더 미리 노출하도록 수정

bunju20 committed:
refactor: 모달 닫기 로직 분리
-utils 폴더에 removeModal 함수 생성
- main.js에 해당 로직 위의 함수로 대체하여 수정

bunju20 committed:
refactor : 유효성 검사 테스트 로직 수정
- validator 코드 수정 및 타입추가
- validator test 코드 수정

bunju20 committed:
refactor : 카테고리, 입력 e2e테스트 코드 수정
- 내장메서드 활용해서 코드 리팩토링

bunju20 committed:
refactor : 기능 테스트 로직 중 입력값 검증 로직 추가

bunju20 committed:
refactor: contain 일부 have.text로 수정하기

bunju20 committed:
refactor: RestaurantList  데이터 분리

bunju20 committed:
refactor : RestaurantItem 컴포넌트 데이터 중심으로 개선

bunju20 committed:
refactor : AddRestaurantModal 데이터 분리

bunju20 committed:
refactor : 레스토랑 init 로직 분리

bunju20 committed:
refactor: 핸들러, 초기로직 분리

bunju20 committed:
docs: 2단계 기능목록 추가

bunju20 committed:
feat: 카테고리 필더 로직 구현

bunju20 committed:
feat: 거리순/이름순 정렬 로직 구현

bunju20 committed:
feat: 자주가는 음식점 UI 구현

bunju20 committed:
docs: 임시커밋

bunju20 committed:
feat: 토글 눌렀을때 자주가는 음식점 용 필터 적용 로직 구현
- 탭을 누르면 드롭다운이 사라진다
- 탭을 눌렀을때 dataset이 변경된다

bunju20 committed:
feat: 레스토랑 아이템 필터링 로직 구현
- 피양콩 할머니를 자주가는음식점으로 설정했을때 필터링 되도록 구현

bunju20 committed:
feat : UI에 빈 별 위치하도록 구현

bunju20 committed:
feat :별을 눌렀을때 상태가 변하는 로직 구현
- 별을 누르면 UI가변하고
- 자주가는 음식점으로 이동하면 별이 눌린 음식점이 보이도록 구현

bunju20 committed:
feat: 레스토랑 터치시 모달창 생기는 로직 구현
- 모달창만 떠오르고 내용은 아직 없음
- 내부의 버튼은 제대로 작동하지 않음

bunju20 committed:
refactor : restaurantList에서 아이템 분리

bunju20 committed:
feat: 모달창에 예시데이터 알맞은 스타일로 출력되는 로직 구현

bunju20 committed:
feat: 누르는 레스토랑의 종류에 따라 모달의 내용 바뀌는 기능 구현

bunju20 committed:
feat: detailmodal 닫기 기능 구현

bunju20 committed:
feat: 레스토랑 삭제 기능 구현

bunju20 committed:
feat: 레스토랑 삭제 localStorage 적용

bunju20 committed:
feat : 즐겨찾기 localStotrage 적용

bunju20 committed:
feat: 음식점 추가 localStorage 적용

bunju20 committed:
fix: 추가한 레스토랑은 안눌리는 오류 수정

bunju20 committed:
fix: 모달창 안에서도 레스토랑이 눌리는 오류 수정

bunju20 committed:
feat: 모달창 배경을 눌러도 꺼지도록 구현

bunju20 committed:
feat: 모달창 안에서 별이 눌리도록 구현

bunju20 committed:
fix: 이벤트 후에 작동안하는 오류 수정
- 탭누른뒤 레스토랑 누르면 안눌리는 오류
- 레스토랑 누른뒤 탭 누르면 안눌리던 오류

bunju20 committed:
fix: 모달 나간후 필터링 안되는 오류 수정

bunju20 committed:
refactor: 레스토랑 타입 ts로 마이그이션

bunju20 committed:
refactor : localStorage ts로 migration

bunju20 committed:
refactor : categoryMapping ts로 migration

bunju20 committed:
refactor : tabHandler ts로 migration

bunju20 committed:
refactor : filterHandler ts로 migration

bunju20 committed:
refactor : favoriteHandler ts로 migration

bunju20 committed:
refactor : detailModalHandler ts로 migration

bunju20 committed:
refactor: restaurantHandler ts로 migration

bunju20 committed:
refactor : modalHandler ts로 migration

bunju20 committed:
feat: DetailModal에서 별 클릭시 RestaurantList의 별 렌더링 되는 기능 구현

bunju20 committed:
docs: README 기능목록 수정

bunju20 committed:
test : 가게 추가 테스트시, 새로생긴 레스토랑을 타겟팅하도록 수정

bunju20 committed:
Merge branch 'bunju20' into step2

onschan commented on src/handlers/detailModalHandler.ts:
핸들러가 어디에서 사용되는지, 어떤 역할을 하는지 예측하기 힘들고 로직들이 비대해지고 있어보여요.
역할에 맞게 쪼개진 컴포넌트들이 본인의 역할에 맞는 동작만 수행하도록 내부에서 이벤트를 관리하는 작업을 우선 진행해보다보면 핸들러의 역할이 다시 보일 수 있을 것 같아요.

bunju20 replied:
코드를 짜는 저도 이게 어디에 사용되는지 혼동되는 부분이 있었어서 이해가 됐습니다ㅎㅎ... 모달 관련 이벤트가 이제 모달 클래스 내부에서 처리되도록 개선해보겠습니다!

onschan commented on src/data/initialRestaurants.ts:
initialRestaurants를 여러 곳에서 사용하고 수정하고 있는 것으로 보이는데,
전역값을 이용시더라도 상태 관리자 모듈을 별도로 두고, CRUD 함수만 거쳐서 상태를 수정하는 구조로 개선하는게 어떨까 싶어요. 현재 구조에서는 여러 곳에서 직접 상태를 수정하기 때문에 데이터 일관성이 깨질 위험이 있어보여요

bunju20 replied:
전역 상태를 수정하고 localStorage에 적용만 했었는데, 확실히 데이터 일관성 문제가 생길것같습니다...! 상태 관리자 모듈을 별도로 구현해서 이부분 개선해보겠습니다!

onschan commented on src/utils/generateId.js:
```diff
@@ -0,0 +1,9 @@
+    0,
+  );
+  return maxId + 1;
+}
```
generateId 함수가 initialRestaurants에 의존하고있네요 🤔
유니크한 아이디를 생성하는 유틸 함수 정도의 레벨이면 충분할 것 같아요.

bunju20 replied:
사용자가 저 한명이라 너무 안일하게 세팅해놓은것같아욥....! id 타입을 string으로 바꾼뒤에 난수랑 타임스탬프 사용해서 해당 함수 수정해보겠습니다!

onschan commented on src/components/RestaurantTabs.js:
```diff
@@ -0,0 +1,22 @@
+
+  container.innerHTML += tabsHTML;
+
+  // 탭 버튼 요소 반환 (이벤트 리스너 추가를 위해)
+  return container.querySelector(".restaurant-tabs");
```
컴포넌트에서 DOM 렌더링 후, 외부에서 리스너를 따로 등록하는 구조로 되어있습니다.
이벤트 리스너까지 컴포넌트 내부에서 처리해, 컴포넌트 자체로 어떤 UI와 어떤 동작을 하는지 예측 가능하도록 처리해주는 편이 좋을 것 같아요.
현재 구조에서는 렌더링마다 리스너가 중복 등록되거나 기능들이 추가될 수록 이벤트 관리가 어려워질 것 같아요.

bunju20 replied:
중복되는듯한 느낌을 받았던 이유가 이런 이유였군요... 이벤트리스너를 컴포넌트 내부에서 처리하도록 개선해보겠습니다. 감사합니다!

onschan commented on src/components/RestaurantItem.js:
```diff
@@ -1,21 +1,25 @@
+  const modalClass = isDetailModal ? " modal-restaurant" : "";
+
+  return `
+    <li class="restaurant${modalClass}"${modalStyle} data-category="${restaurant.category}" data-favorites="${restaurant.favorites}" data-restaurant-id="${restaurant.id}">
```
data 속성에 꽤 많은 정보들이 들어가고있는데요. 현재 관리하고 있는 상태로 충분하지는 않았는지 궁금하네요.

bunju20 replied:
그부분을 생각하지 못하고 개발한것 같습니다. DOM을 데이터 저장소처럼 써버린 느낌이네욥...  id는 그대로 유지하고, 나머지는 저장하는 곳에서 ID로 조회하면 좋을것같습니다!  이부분도 개선해보겠습니다!

onschan commented on src/components/AddDetailModal.js:
```diff
@@ -0,0 +1,24 @@
+import CustomButton from "../shared/CustomButton.js";
+import RestaurantItem from "./RestaurantItem.js";
+
+export function AddDetailModal(container, selectedRestaurant) {
```
컴포넌트 레벨로 선언되어있지만 반환값이 존재하지 않고 dom 조작을 실행하기만하는 역할을 하고 있어보여요. 함수형 컴포넌트가 자체로 UI와 로직이 있고 해당 내용을 반환하도록 개선이 되면 좋을 것 같아요.

bunju20 replied:
위에서 피드백을 보고 해당부분 리팩토링 진행하고있습니다!
UI생성과 이벤트 핸들링을 최대한 내부적으로 관리하도록 수정해봤고, 다른부분도 같은 방향으로 개선해보겠습니다!

onschan commented on src/components/RestaurantItem.js:
```diff
@@ -1,21 +1,25 @@
 
-  container.innerHTML += `
-    <li class="restaurant">
+export default function RestaurantItem(restaurant, isDetailModal = false) {
```
다른 컴포넌트와 다르게 해당 컴포넌트는 반환값을 가지고 있는데요. 
전체적인 컴포넌트의 개선 방향이 이 RestaurantItem 과 같은 형식으로 관리가 되면 어떨까싶어요. 

내부에서 이벤트 리스너 등록과 함께 필요한 UI를 반환하고 이 컴포넌트들을 바탕으로 로직들을 조합해가는 과정을 겪으셨으면 좋곘어요

bunju20 replied:
DOM 조작 대신에 render 메서드로 UI를 반환하게 하고, 필요한 이벤트 리스너를 내부에서 설정하게 (AddRestaurantModal, RestaurantFilterContainer, RestaurantList )수정하겠습니다!

onschan commented on src/data/initialRestaurants.ts:
```diff
@@ -0,0 +1,64 @@
+  },
+];
+
+export const initialRestaurants:Restaurant[] = initializeRestaurants(defaultRestaurants);
```
initializeRestaurants는 `localStorage에서 레스토랑 데이터 초기화하기`라는 역할을 지니고 있는 걸로 확인했는데요.
데이터를 관리하고 반환하는 역할을하고 있는 해당 레벨에서 관리 하는것이 전체 어플리케이션 레벨에서 예측이 가능할까요?

bunju20 replied:
initializeRestaurants가 파일 불러올때 자동으로 실행돼서 데이터 초기화가 어디서 되는지 예측하기 어려웠던것같습니다! 애플리케이션 시작 시점(main.js)에서 명시적으로 초기화할수 있게 한번 수정해보겠습니다!

onschan commented on src/handlers/detailModalHandler.ts:
```diff
@@ -0,0 +1,162 @@
+import { Restaurant } from "../../types/Restaurant.ts";
+
+let selectedRestaurantId : number | null = null;
+let eventListenersAttached : boolean = false;
```
selectedRestaurantId는 상태로 보이는데 컴포넌트 위계에 맞게 잘 관리되고 있어보이지는 않아요.

컴포넌트를 함수로 만드셨어도 UI와 밀접한 상태는 관리하고있어야하며 각 컴포넌트가 인지와 반환값으로 역할에 맞게 동작해야돼요.

bunju20 replied:
넵! 레스토랑 ID 상태가 컴포넌트 외부에서 관리되고 있었는데, 컴포넌트 내부로 이동시켜서 개선해보겠습니다. 감사합니다!

onschan commented on src/handlers/detailModalHandler.ts:
```diff
@@ -0,0 +1,162 @@
+import { Restaurant } from "../../types/Restaurant.ts";
+
+let selectedRestaurantId : number | null = null;
+let eventListenersAttached : boolean = false;
```
eventListenersAttached도 결국 이벤트가 매끄럽게 관리되지 않고 있어서 생긴 변수로 생각되는데,

컴포넌트들을 인자와 반환값과 자기 자신에게 부여된 역할을 내부의 이벤트 핸들러를 통해 수행하는 것으로 먼저 구조 개선 후에 다시 살펴봐도 좋을 것 같아요

bunju20 replied:
넵 이부분 이벤트 리스너 설정 로직을 DetailModal 클래스의 정적 메서드로 이동시키고,  외부 상태 변수 없이 관리될수 있게 개선해봤습니다...! 생각해보니 뭐가 안되면 구조를 살펴봤어야 했는데, 코드를 계속 추가했던것 같습니다😔

onschan commented on src/handlers/detailModalHandler.ts:
```diff
@@ -0,0 +1,162 @@
+  setupRestaurantItemEventListeners();
+  setupFilterEventListeners();
+  setupFavoriteEventListeners();
+}
```
다른 핸들러에서 선언된 로직들과 결합도가 높고 예측이 힘들지 않을까 생각이 드네요 😅

bunju20 replied:
DetailModal 클래스가 자기의 이벤트 처리를 스스로 관리할수있게 한번 개선해봤습니닷! 이벤트 리스너를 한꺼번에 재설정하는 대신, 필요한 시점에 필요한 리스너만 설정할수있게 했어야 했던것 같아요

onschan commented on src/handlers/restaurantHandler.ts:
```diff
@@ -0,0 +1,105 @@
+  setupFavoriteEventListeners();
+
+
+}
```
기본 프리티어에 eol 설정 해주시는것도 유용해요!

bunju20 replied:
아앗 말씀해주셔서 감사합니다! 설정해놓겠습니다☺️ !!

onschan commented on types/Restaurant.ts:
```diff
@@ -0,0 +1,12 @@
+  name: string;
+  category: Category;
+  categoryName: CategoryName;
+  distance: string;
```
string 대신 number 형으로 선언하고 이용하는게 훨씬 용이할 것 같아요.
이미 여러곳에서 사용되고있고 수정이 어렵다면 반영은 안해주셔도됩니다.

bunju20 replied:
number형이 훨씬 좋을것같습니다. 별생각없이 string으로 구현하고, 정렬할때는 일일히 Number로 변환해서 사용했는데요... 수정하겠습니다!

onschan reviewed (changes requested):
안녕하세요. 분주

미션 진행해주시느라 수고 많으셨어요.

앱이 잘 동작되도록 구현을 잘 해주셨지만 컴포넌트와 상태에 대한 개선이 필요할 것 같아요.

1단계에서는 로직이 단순하고 책임이 이렇게 파생되지 않아서 문제가 발생하지 않았을지 몰라도 지금의 단계에서는 구조적 개선이 필요할 것 같아요.

단순히 innerHTML만 실행하는 컴포넌트들과 그에 따라 파생되고있는 이벤트 핸들러 로직들을 먼저 필요한 element를 반환하고 본인의 리스너를 갖는 함수 자체가 해당 UI와 로직을 나타내는 컴포넌트로 동작할 수 있도록 수정해주시면 좋을 것 같아요. 

---
>이벤트 리스너 관리
하면서 이벤트버블링과 이벤트리스너 누락으로 인한 오류가 잦았습니다. 예를 들면 레스토랑을 눌러서 상세모달이 띄워진후에는 특정 버튼들의 이벤트 리스너가 없어졌다던지... 같은문제로요! 그래서 매번 리스너를 다시 붙여주는 식으로 구현을 했는데 이게 맞는지 잘 모르겠습니다. 비효율적이지 않나?? 하는 생각을 하면서 구현한것 같아요. 좀더 좋은 방식이 있나요...!?

돔을 조작하는 함수들과 핸들러들이 너무 멀리서 관리되고 예측하기 힘들도록 관리되고 있는게 어느정도 영향을 주지 않았을까 싶어요.

컴포넌트 단위에서 이벤트를 바라봐 주시고 각 컴포넌트들마다 이벤트를 관리하는게 용이하지 않거나 비효율적인 경우 부모 레벨에서 이벤트 위임을 활용하는 방식도 차용해보면 좋을 것 같아요.

이벤트에 대한 관리 책임은 각 컴포넌트들에 있도록 개선이 되면 방향성이 다시 보일 수 있을 거라고 생각해요.

>함수형 vs 클래스 기반 설계
사실 함수형으로 프로젝트를 진행해본게 이번이 처음인데, 개발하면서 불편함을 많이 느낀것같습니다! 상태 관리와 이벤트 핸들러 관리 측면에서 클래스 기반으로 설계했다면 더 나았을지 궁금합니다. 현재 구조에서는 전역 변수와 함수들이 여러 파일에 분산되어 있다고 느껴졌던적이 많았던것같아요!

이전에 컴포넌트에 대해 다시 한번 생각해주시면 좋을것 같아요. 

RestaurantItem 외에 다른 컴포넌트들은 단순히 돔을 조작하는 함수처럼 동작하고 있어요.

인자와 반환값을 바탕으로 필요한 UI와 로직들이 캡슐화되어있는 조각 이라고 생각하고 필요한 이벤트에 대한 책임도 컴포넌트가 들고 있게끔해주세요.

내부 상태를 유지하는 방법등 함수와 클래스가 다른 점이 있지만 컴포넌트라는 역할이라는 핵심 개념은 동일합니다.

>DOM 조작과 상태 동기화
UI 상태와 데이터 상태를 동기화하는 부분에서도 어려움이 있었는데요... 예를 들어 즐겨찾기를 모달에서 변경하면 목록에는 반영되지 않는 문제가 있어서 즐겨찾기 버튼 클릭 이벤트로 레스토랑 목록을 다시 렌더링하는식으로 구현했는데... 너무 비효율적인것처럼 느껴졌습니다! 더 좋은 방법이있을까요??

상태가 모호하게 관리되고 있는 부분이 문제이지 않을까해요. 

현재 식당 목록에 대한 상태를 어떻게 관리하고 있는지 일관성있는 흐름에 따라 정리를 진행하고 이벤트 트리거마다 setup~ 등의 호출보다는 해당 상태가 변경되었을때 필요한 컴포넌트를 다시 렌더링하는 방향으로 진행하는 것이 어떨까 싶어요.

bunju20 committed:
refactor: DetailModal 컴포넌트화 및 이벤트 핸들러 책임 이전

- AddDetailModal을 DetailModal 클래스로 리팩터링
- 모달관련 이벤트 핸들러를 컴포넌트 내부로 이동
- 이벤트 리스너 관리 개선

bunju20 committed:
Merge branch 'step2' of https://github.com/bunju20/javascript-lunch into step2

bunju20 committed:
refactor : 중앙 상태 관리 로직 구현
- initialRestorants 전역변수대신에 restaurantStore 모듈 구현
-데이터 접근 및 수정 로직 일관성 개선

bunju20 committed:
refactor: generateId 생성 로직 변경 및 restaurant type 수정

bunju20 committed:
refactor: restaurantTab 이벤트리스너 구조 개선
- handler의 내용을 컴포넌트로 옮기고 리팩토링 하였음
- 잘작동하는지 테스트 완료 하였음

bunju20 committed:
refactor : data 속성 정보 restaurantStore로 관리하도록 수정

bunju20 committed:
refactor : DetailModal의 핸들러 DetailModal class로 이동

bunju20 committed:
refactor : addRestaurantModal 로직 수정

- DOM조작 분리
- 내부에서 이벤트 리스너 설정

bunju20 committed:
refactor : RestaurantFilter 구조 수정

- DOM 로직 분리

bunju20 committed:
refactor: RestaurntList 구조 수정
-DOM로직 분리
-초기화 코드에서 해당 구조 사용하도록 수정

bunju20 committed:
refactor: RestaurntTab 구조 수정
- DOM 분리

bunju20 committed:
refactor: rerender 로직 컴포넌트로 이동

bunju20 committed:
refactor : 애플리케이션 시작 시점에 명시적으로 초기화하도록 수정
- localStorage 저장과 불러오기만 남김
- restaurantStore 초기화 메서드로 따로 분리함
- main.js initialize를 시작지점에 호출함

bunju20 committed:
refactor: detailModal 이벤트 핸들러 클래스 내부로 이동

bunju20 committed:
refactor : Restaurant의 distance를 number형으로 수정

onschan commented on src/components/DetailModal.js:
```diff
@@ -0,0 +1,90 @@
+      existingModal.remove();
+    }
+
+    this.container.insertAdjacentHTML("beforeend", modalHTML);
```
insertAdjacentHTML 사용 좋아요 😁

onschan commented on src/components/RestaurantList.js:
```diff
@@ -1,13 +1,10 @@
 export default function RestaurantList(
   container,
-  restaurants = initialRestaurants,
+  restaurants = restaurantStore.getRestaurants(),
```
👍
전역 스토어 개념으로 잘 사용해주셨는데,
바닐라 JS 기반 미션을 다시 할 기회가 있다면, 전역 스토어 사용없이 구현해보는 과정을 먼저 겪었으면 어떨까 싶기도 해요. 
리뷰를 하다보니 사용할 상태를 어디서 어떻게 관리해야되고 상태가 어디까지 어떤식으로 전파되어야되고 상태 변경의 책임과 권한은 누가 가지고 있어야되는지 고민할 포인트들을 많은 크루들이 전역 스토어의 사용으로 놓치고 넘어가는 것 같아서 아쉽다고 느꼈어요.
레벨 2로 넘어가기 전에 한번쯤 고민해보고 넘어가는 것도 좋을 것 같아요.

onschan commented on src/store/restaurantStore.ts:
```diff
@@ -61,4 +129,4 @@ const defaultRestaurants: Restaurant[] = [
 ];
 
-export const initialRestaurants:Restaurant[] = initializeRestaurants(defaultRestaurants);
+export const restaurantStore = new RestaurantStore(defaultRestaurants);
```
전역 싱글톤은 크게 추천드리지는 않고 싶어요.
어플리케이션 레벨을 벗어나있는 어디서든 참조가 가능하다는건 관리 측면에서 신뢰성이 너무 낮아진다고 생각해요.
식당 정보를 전역으로 관리할만한 데이터 일까 싶기도하고요.

변경하시려다보면 변경 지점이 많아질 것으로 보여서 반영은 안해주셔도 돼요. 다만 이렇게 쉽게 사용하는 상태가 학습 관점에서 어떤 고민의 포인트를 놓치게하고 있는지는 생각해주세요!

onschan commented on src/utils/generateId.js:
```diff
@@ -1,7 +1,5 @@
 import { restaurantStore } from "../store/restaurantStore.ts";
```
generateId는 restaurant를 모르게 하는김에 깔끔하게 요 부분도 제거해버리죠.

onschan commented on src/components/RestaurantTabs.js:
```diff
@@ -1,22 +1,106 @@
+    this.tabsElement = null;
+    this.tabChangeListeners = [];
+    this.setupInitialState();
+  }
```
헤더의 책임을 갖는 상태들을 선언하고 관리하는 부분으로 개선이 되었네요.
의도드린 부분들을 잘 파악하시고 개선해주셨어요. 감사합니다 😁

이 경험을 통해 "상태"가 선언되는 위치와 전파 범위 등 상태 관리에 대해 더 명확히 경험해보셨으면 좋어요!

onschan reviewed (approved):
안녕하세요 분주!

개선 작업을 진행하시면서 어떤 생각을 하셨을지 궁금하네요.
효용을 느끼셨을까요? 아니면 또 다른 한계들이 발견되었을까요?

저는 두가지 모두 느끼셨으면 좋겠어요 😁
지금의 구조가 최고의 구조가 아닐거에요.
지속적으로 한계를 맞이하고 그를 통해 구조적인 개선에 대해 고민하고 해결하는 과정을 통해 타협하기도하고 기준을 세우기도하면서 프로젝트에 적합한 방법과 구조가 점점 탄탄해지게될거에요.
그렇게 구성한 설계와 구조가 너무 이쁘고 아름답다면 추상화하며 프레임워크를 만들어 버릴 수도 있고요 👀

미션 잘 수행해주셨고 앞으로도 더 많이 고민하시고 그를 통해 더 성장하셨으면 좋겠어요!
이번 미션은 approve 하도록 할게요! 수고하셨어요.
